File: scripts/reference-solutions/network-upgrade-routing.cpp
#include <iostream>
#include <queue>
#include <tuple>
#include <vector>
using namespace std;

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);

  int n, m, k;
  cin >> n >> m >> k;
  vector<vector<pair<int, long long>>> adj(n + 1);
  for (int i = 0; i < m; ++i) {
    int u, v;
    long long w;
    cin >> u >> v >> w;
    adj[u].push_back({v, w});
  }

  const long long INF = (long long)4e18;
  // dist[node][used] = min latency to reach `node` having spent `used` upgrade
  // credits. Dijkstra over these (node, used) states.
  vector<vector<long long>> dist(n + 1, vector<long long>(k + 1, INF));
  dist[1][0] = 0;

  using State = tuple<long long, int, int>;  // (latency, node, used)
  priority_queue<State, vector<State>, greater<State>> pq;
  pq.push({0, 1, 0});
  while (!pq.empty()) {
    auto [d, u, used] = pq.top();
    pq.pop();
    if (d > dist[u][used]) {
      continue;
    }
    for (auto& [v, w] : adj[u]) {
      long long nd = d + w;
      if (nd < dist[v][used]) {
        dist[v][used] = nd;
        pq.push({nd, v, used});
      }
      if (used < k && d < dist[v][used + 1]) {
        dist[v][used + 1] = d;
        pq.push({d, v, used + 1});
      }
    }
  }

  long long best = INF;
  for (int used = 0; used <= k; ++used) {
    best = min(best, dist[n][used]);
  }
  cout << (best != INF ? best : -1) << '\n';
  return 0;
}

File: scripts/reference-solutions/network-upgrade-routing.py
import sys
import heapq


def main():
    data = sys.stdin.buffer.read().split()
    idx = 0
    n = int(data[idx])
    m = int(data[idx + 1])
    k = int(data[idx + 2])
    idx += 3

    adj = [[] for _ in range(n + 1)]
    for _ in range(m):
        u = int(data[idx])
        v = int(data[idx + 1])
        w = int(data[idx + 2])
        idx += 3
        adj[u].append((v, w))

    INF = float("inf")
    # dist[node][used] = min latency to reach `node` having spent `used` upgrade
    # credits. Dijkstra over these (node, used) states.
    dist = [[INF] * (k + 1) for _ in range(n + 1)]
    dist[1][0] = 0
    pq = [(0, 1, 0)]
    while pq:
        d, u, used = heapq.heappop(pq)
        if d > dist[u][used]:
            continue
        for v, w in adj[u]:
            # Pay the latency.
            nd = d + w
            if nd < dist[v][used]:
                dist[v][used] = nd
                heapq.heappush(pq, (nd, v, used))
            # Spend a credit to make this link free.
            if used < k and d < dist[v][used + 1]:
                dist[v][used + 1] = d
                heapq.heappush(pq, (d, v, used + 1))

    best = min(dist[n])
    print(best if best != INF else -1)


main()
